server/polar/order/service.py

```diff
@@ -751,6 +773,164 @@ async def trigger_payment(
+        self, session: AsyncSession, order: Order, confirmation_token_id: str
+    ) -> CustomerOrderPaymentConfirmation:
+        """
+        Process retry payment with direct confirmation (confirm=True).
+        Follows checkout flow pattern - creates PaymentIntent and lets webhooks handle everything else.
+        """
+
+        if order.status != OrderStatus.pending:
+            raise OrderNotPending(order)
+
+        if order.next_payment_attempt_at is None:
+            raise OrderNotEligibleForRetry(order)
+
+        if order.subscription is None:
+            raise OrderNotEligibleForRetry(order)
+
+        if order.payment_lock_acquired_at is not None:
+            log.warning(
+                "Payment already in progress",
+                order_id=order.id,
+                lock_acquired_at=order.payment_lock_acquired_at,
+            )
+            raise PaymentAlreadyInProgress(order)
+
+        stripe_customer_id = order.customer.stripe_customer_id
+        if stripe_customer_id is None:
+            raise MissingOrderStripeCustomerID(order)
+
+        customer_repository = CustomerRepository.from_session(session)
+        customer = await customer_repository.get_by_id(order.customer_id)
+        assert customer is not None, "Customer must exist"
+
+        org_repository = OrganizationRepository.from_session(session)
+        organization = await org_repository.get_by_id(customer.organization_id)
+        assert organization is not None, "Organization must exist"
+
+        # Create payment intent metadata (similar to checkout flow)
+        metadata: dict[str, Any] = {
+            "order_id": str(order.id),
+        }
+        if order.tax_rate is not None:
+            metadata["tax_amount"] = str(order.tax_amount)
+            metadata["tax_country"] = order.tax_rate["country"]
+            metadata["tax_state"] = order.tax_rate["state"]
+
+        try:
+            order_repository = OrderRepository.from_session(session)
+            lock_acquired = await order_repository.acquire_payment_lock(order.id)
+            if not lock_acquired:
+                log.warn("Payment already in progress", order_id=order.id)
+                raise PaymentAlreadyInProgress(order)
+
+            payment_intent = await stripe_service.create_payment_intent(
+                amount=order.total_amount,
+                currency=order.currency,
+                automatic_payment_methods={"enabled": True},
+                confirm=True,
+                confirmation_token=confirmation_token_id,
+                customer=stripe_customer_id,
+                statement_descriptor_suffix=organization.statement_descriptor,
+                description=f"{organization.name} — {order.product.name}",
+                metadata=metadata,
+                return_url=settings.generate_frontend_url(
+                    f"/portal/orders/{str(order.id)}"
+                ),
+            )
+
+            if payment_intent.status == "succeeded":
```

Comment:
Question: we immediately have a "final" status after calling this? There is no kind of "pending" state where we should wait for a webhook or something?

Reply:
Discussing below

Reply:
Where? 😊

Reply:
Sorry, I confused it! If I look correctly at our Stripe Dashboard we only accept payments with immediate confirmation, so no bank debits. Therefore there is no "pending" status. The "required_action" it's the only thing that we should consider, and we should wait for the webhook to get it confirmed.